designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,308 @@
+- `@typescript-eslint/eslint-plugin` has its own [`applyDefault`](https://github.com/typescript-eslint/typescript-eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/packages/utils/src/eslint-utils/applyDefault.ts#L10) used in a wrapping [`createRule`](https://github.com/typescript-eslint/typescript-eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/packages/utils/src/eslint-utils/RuleCreator.ts#L84)
+
+Although the currently used Ajv package does provide a [`useDefaults` option](https://ajv.js.org/guide/modifying-data.html#assigning-defaults), rule developers typically have not used it.
+Ajv does not recursively fill in objects for rule defaults - an ergonomics issue for rule authors.
+
+In addition to increasing development complexity for rules, the lack of ergonomic options defaulting means end-users haven't had a consistent or programmatic way to determine rule defaults.
+The only user-facing method to determine a rule's default options is to read the documentation, or if it either doesn't exist or doesn't describe default values, read the rule's source.
+Even if a plugin does have documentation, there is no guaranteed consistency in phrasing.
+
+For example:
+
+- [`array-callback-return`](https://eslint.org/docs/latest/rules/array-callback-return#options) phrases its options as _`"<key>": <value> (default) <explanation>`_
+- [`accessor-pairs`](https://eslint.org/docs/latest/rules/accessor-pairs#options) phrases its options as _`<key> <explanation> (Default <default>)`_
+
+This RFC proposes adding a `defaultOptions` property to rules' `meta`.
+Doing so would:
+
+- Streamline the process of creating and maintaining rules that take in options
+- Standardize how both developers and end-users can reason about default rule options
+- Encourage writing rules that allow tooling such as [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator) to describe rule options programmatically - and therefore more consistently
+
+Options provided to rules would be the result of a deep (recursive) clone and merge of the rule's default options and user-provided options.
+
+## Detailed Design
+
+Currently, ESLint rule options are passed through a [`getRuleOptions` function](https://github.com/eslint/eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/lib/linter/linter.js#L740) whose only processing is to remove severity.
+That function could be augmented to also take in the `rule.meta.defaultOptions`, if it exists.
+
+Defaulting logic could go with the following logic to approximate user intent:
+
+1. If the user-provided value is `undefined`, go with the default option
+2. If the user-provided value is an object, a new `{}` object is created with values recursively merged
```

Comment:
What if the rule schema has different object schemas in oneOf/anyOf, for example as in [prefer-destructuring](https://github.com/eslint/eslint/blob/994596b07f5ff20a615a4be1ea03e5fd59cdb84b/lib/rules/prefer-destructuring.js#L43-L86). The default for this would be an [object](https://github.com/eslint/eslint/blob/994596b07f5ff20a615a4be1ea03e5fd59cdb84b/lib/rules/prefer-destructuring.js#L107-L110) with `VariableDeclarator` and `AssignmentExpression` properties, but when user config provides the other alternative with `array` and `object` properties, merging would result in an object with a mix of properties that would fail validation. Is this a case where the rule should not use `meta.defaultOptions`?

Reply:
In general the advice we give when using the current ts-eslint tooling is that if you've got some complex options schema - eg oneOf where the schemas are truly disjoint like your example, or rules like [`@typescript-eslint/ban-types`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/ban-types.ts) - they are just usecases where the standard logic cannot apply and thus you need to write your own options merging logic.

Reply:
I think at this point we should be discouraging the use of `oneOf` and `anyOf`. We ended up doing this in core rules because we didn't realize just how many exceptions and options people would want on rules, so early rules would often just use a string value as an option. In order to allow more options, we'd also allow an object. In general, I think we should be encouraging folks to always use an object with multiple properties for their rule options and avoid `oneOf` and `anyOf` in their schemas.

Reply:
> I think we should be encouraging folks to always use an object with multiple properties for their rule options and avoid oneOf and anyOf in their schemas.

Aha! https://github.com/typescript-eslint/typescript-eslint/discussions/6040. I would be interested in filing an issue/discussion on ESLint core to discuss establishing this in core development docs.

Reply:
Even with that as a constraint, i'd still have many use cases for anyOf/oneOf, because there are often multiple disjoint sets of options that may or may not be mutually exclusive.

Reply:
@ljharb you can always ignore recommendations. I just think it's helpful to give people guardrails that cover the 80% case. If you fall into the 20% case, then void the warranty and do what you need to. I think it's fine for the 20% case to not get the 80% case DX.

Reply:
Can you provide some examples?

Reply:
@JoshuaKGoldberg for example, https://github.com/jsx-eslint/eslint-plugin-react/blob/9da1bb0f4aa44849b56e8c2064afd582ea8b36f8/lib/rules/jsx-indent-props.js#L64-L71 - that object property takes a number or two enum values.

https://github.com/jsx-eslint/eslint-plugin-react/blob/9da1bb0f4aa44849b56e8c2064afd582ea8b36f8/lib/rules/jsx-max-props-per-line.js#L39-L70 is more complex; you can specify a maximum with single and/or multi, xor you can specify a maximum with "when".

There's a bunch more in this plugin, and I haven't looked at the other two yet, but suffice to say it's pretty much impossible to avoid with a complex rule that evolves over time.